src/OmniSharp.Host/Services/DotNetCliService.cs

```diff
@@ -98,10 +99,6 @@ public Task RestoreAsync(string workingDirectory, string arguments = null, Actio
                     }
                     finally
                     {
-                        _semaphore.Release();
-
-                        _locks.TryRemove(workingDirectory, out _);
```

Comment:
Would it not be simpler to just remove this line and force all restores to the same WorkingDirectory to share the same lock object?

Reply:
Sharing the same lock object would mean keeping a dictionary of semaphoreslim. But then you need to clean up when you're ready otherwise the semaphoreslim remains in memory. That's where this library comes in handy.

Reply:
So, just so I am sure, you are trying to make restore operations run in parallel as long as they are in different working directories?

Reply:
Yes, the way I understand it is that they currently already run in parallel, irrespective of working directory.

Reply:
Stepping back and looking at the restore code. I think ideally, we wouldn't be passing in a working directory. Instead, we should be passing in the project path. Reason being that a directory could contain a mix of project and solution files which will cause the dotnet restore command to fail. Of course, then we would want to ensure we only start one Restore operation at a time for the project path.

Reply:
@JoeRobich so basically you'd approve this PR but would want improvements? I don't want to spend more time if you then have an issue with the 3rd party library.

Reply:
Yeah, sorry, wish we could have discussed this in an issue first. I think the effort would be better spent getting the restore code to do the right thing. Thanks for bringing this to my attention.

Reply:
@JoeRobich I'm sorry but your answer is a bit ambiguous. Would you like me to make changes to my PR to pass on the projectPath instead of the workingDirectory?

Reply:
@MarkCiliaVincenti I don't want to take on a new dependency for this fix. I think the best course of action would be to use a ConcurrentDictionary (or something similar) to track restore tasks running for a particular path and avoid locking entirely. If you want to work on a solution like that, I would be happy to take it.

Reply:
@JoeRobich if you don't want to take the library I suggest you go for AsyncDuplicateLock from https://stackoverflow.com/a/31194647/9945524

This will not be as performant as the AsyncKeyedLock library and also has much more memory allocations. Moreover you're stuck with code you'll have to maintain in future.

Reply:
@JoeRobich shall we close this PR?